### PR TITLE
Dropout (continued)

### PR DESCRIPTION
**Pull request by gbaydin**
This is to keep track of the work on dropout, batchnorm, and some related operations needed for machine learning.

It follows from the previously merged PR #112.

### CODE REVIEW

**Comment by gbaydin** [comment]
Merging this to dev now. This PR has made significant progress in several areas:
- bernoulli, multinomial distributions
- dropout operations and nn layers
- nn train/eval mode
- support indexing dimensions backwards in some ops
- clamp operation
- core differentiable distributions (currently uniform, normal, bernoulli, categorical) and an empirical distribution that provide a solid base for probabilistic programming systems
- a few other improvements and fixes

I intend to continue with a general transpose(dim1, dim2) operation in a separate PR. This is one of the ingredients needed to implement batch normalization layers.
